src/mono/wasm/runtime/hybrid-globalization/change-case.ts

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { Module } from "../imports";
+import { setU16 } from "../memory";
+import { mono_wasm_new_external_root } from "../roots";
+import { conv_string_root } from "../strings";
+import { MonoString, MonoStringRef } from "../types";
+import { Int32Ptr } from "../types/emscripten";
+import { pass_exception_details } from "./common";
+
+export function mono_wasm_change_case_invariant(exceptionMessage: Int32Ptr, src: number, srcLength: number, dst: number, dstLength: number, toUpper: number) : void{
+    try{
+        const input = get_utf16_string(src, srcLength);
+        let result = toUpper ? input.toUpperCase() : input.toLowerCase();
+        // Unicode defines some codepoints which expand into multiple codepoints,
+        // originally we do not support this expansion
+        if (result.length > dstLength)
+            result = input;
+
+        for (let i = 0; i < result.length; i++)
```

Comment:
Since these loops seem to be hot in your measurements, it might be worth optimizing them slightly like this:
```javascript
for (let i = 0, j = dst; i < result.length; i++, j += 2)
  setU16_unchecked(j, result.charCodeAt(i));
```
If setU16_unchecked doesn't exist it's worth adding, it can work like the other unchecked setters. We know charCodeAt's return value will always be in range (MDN says the return value is always 0-65535, unless index is out of range)

Reply:
It does not have a visible impact on the time when tested with 64kB string but it looks like a good idea, so I am applying it (to https://github.com/dotnet/runtime/pull/85516, here we have normalization changes only).

Reply:
Yeah it's possible both JS runtimes optimize it already, value range analysis is done in them I think so it's possible they already know that setU16's check can never fail. The `i * 2` might also get hoisted out of the loop or just be equivalent-cost to the `j += 2` due to how arithmetic works in JS...